### DISCUSSION TITLE
(user error, resolved)

### DISCUSSION THREAD

**Discussion by kingpalethe** [openzfs/zfs#16571] [General]
I previously was able to install ZFS on 22.04, and 24.04, and it seemed to work fine -- but recently if I do any software update on either of this LTS releases, ZFS fails completely -- when I try to run it, I get the `Failed to initialize the libzfs library` error. I think Canonical may have recently pushed an update which kills ZFS. 

I've made multiple attempts to reinstall ZFS but have not yet succeeded. 

As far as I can tell these directions are out of date: https://openzfs.github.io/openzfs-docs/Getting%20Started/Debian/index.html
 -- I get a lot of error messages about PGP keys.

Previously I had used...
```
sudo apt install zfsutils-linux
````

But this installs version ZFS 2.2.2. It was working previously, but now that, after I reboot I get the `Failed to initialize the libzfs library` error, I realize that 2.2.2 is very old and  not compatible Linux 6.8 kernel.

It also seems that Canonical have modified their installers going back to 22.04, such that it's not straightforward to install a version of Ubuntu that has an old enough kernel to be compatible with ZFS 2.2.2. 

So -- I should try building from source, right?

I've tried starting here: https://zfsonlinux.org/ ... and starting from here: https://openzfs.github.io/openzfs-docs/Developer%20Resources/Building%20ZFS.html#configure-and-build .... all of these fail with multiple errors like

```
r rpc/xdr.h... no
configure: WARNING: cannot find tirpc via pkg-config or in the standard locations
```

I've also tried going to the "Ubuntu" installation instructions here: https://openzfs.github.io/openzfs-docs/Getting%20Started/Ubuntu/index.html

Which seem to push me back to 2.2.2. 

Can anyone give me any hints for the best strategy to get ZFS working under Ubuntu?

I don't want to install it on my root boot disk or anything fancy -- I just want to set up a zpool mirror...

thanks!

**Comment by kingpalethe** [comment]
One thought I -- recently started adjusting the size of my ARC memory....  which I do by..

running this command....
```
sudo nano /etc/modprobe.d/zfs.conf
```
adding this line to zfs.conf .....
```
options zfs zfs_arc_max=options zfs zfs_arc_max=8589934592
```
and then running this command...

```
sudo update-initramfs -u
```
this last command produces the output 
`update-initramfs: Generating /boot/initrd.img-6.8.0-45-generic`

After I do this, I reboot, and ZFS fails to load with the `Failed to initialize the libzfs library` error. 

Should I just give up on this and assume there is no safe way to adjust any ZFS settings like this under Ubuntu? 

**DISCOVERY:
As long as I never run this command `sudo update-initramfs -u` -- it seems I am still able to run ZFS, even though it is version 2.2.2 and this should be incompatible with Linux kernel 6.8 ... I'm really not sure how stable it is, but at least if I avoid this `sudo update-initramfs -u` , then ZFS does seem to work...**

**Comment by rincebrain** [comment]
[See here](https://openzfs.github.io/openzfs-docs/Developer%20Resources/Building%20ZFS.html#install) for how Ubuntu broke that a long time ago.

**Comment by kingpalethe** [comment]
Thanks, are you sure that is the right link? That seems to be the main OpenZFS directions, which I found were outdated. Do you have any recommended strategy to install ZFS on Ubuntu?

**Reply by rincebrain** [reply]
> On Ubuntu, for example, the modules from this repository install in the extra kernel module path, which is not in the standard depmod search path. Therefore, for the duration of your testing, edit /etc/depmod.d/ubuntu.conf and add extra to the beginning of the search path.

The instructions aren't outdated, you just skimmed them.

**Comment by kingpalethe** [comment]
Here is some relevant discussion. https://bugs.launchpad.net/ubuntu/+source/zfs-linux/+bug/2044657

From what I can understand, Canonical considers that 2.2.2 should still work with modern Ubuntu.

**Reply by rincebrain** [reply]
I generally recommend using any distro except Ubuntu for ZFS, but that's just my personal opinion, not any sort of "official" project take.

They tend to break things and not notice or fix them for months in their packages, ship patches to their ZFS that eat your data, have bugs in their ZFS management daemon that eat your data, and at least once, they had a footnote in their dist-upgrade notes of "don't dist-upgrade if you're on ZFS root, it will fail to boot".

**Reply by kingpalethe** [reply]
Thanks. Can you recommend the most n00b-friendly distribution that you think provides the cleanest and easiest ZFS experience? I realize this is a matter of opinion.. but I assume this discussion could useful for others as there are likely other people surprised (like I have been) that ZFS on Ubuntu is such a mess, and probably need to jump ship to a different distro.

**Reply by kingpalethe** [reply]
For some background - I do need a GUI and the ability to install other mainstream software like VS-Code. But the upmost priority is setting up an installation that can be "left alone" and keep running for many months (ideally years) without reboot or fiddling around with stuff.

**Reply by rincebrain** [reply]
I like Debian, but that's because I know where the warts are.

I generally recommend against Fedora and similar fast-moving distros, because there's going to be a timelag invariably between kernel releases and support for them, and Fedora/Arch/etc ship new kernels fast enough that this is going to break you sometimes, unless you have the budget to pay someone to fix kernel breakages. ;)

RHEL-alikes are the ones that I would describe as "most well-tested by the maintainers", since that's what I believe LLNL and friends are using, but various people dislike how old the software on those tends to be.

That said, I would suggest you report a problem to Ubuntu, if you were just using purely Ubuntu packages and things broke - I expect many levels of failure, but "ZFS doesn't run at all" is a novel level of breakage.

**Reply by kingpalethe** [reply]
Thanks much. And you think if I installed Debian (bookworm) with Gnome, then I could install ZFS using these directions? https://openzfs.github.io/openzfs-docs/Getting%20Started/Debian/index.html

The Ubuntu directions on that website (https://openzfs.github.io/openzfs-docs/Getting%20Started/Ubuntu/index.html ) , as far as I can tell, are out of date and no longer work. So that's why I ask. I really would rather rely on packages instead of trying to build stuff from source.

**Reply by rincebrain** [reply]
The Ubuntu instructions worked for me when I tried them just now in a 24.04 VM.

I would suggest you look at your installed packages, something like `dpkg -l | grep -E '(zfs|zpool|nvpair)'`. I would guess you have either a mix of packages, or stuff in /usr/local/ and packages, breaking it, offhand.

**Reply by kingpalethe** [reply]
Thanks.

 I don't really have a reproducible test case unfortunately. But maybe this discussion will be found by other people who have this problem and they'll chime in.  

My current understanding of the issue is......

1) I added this line `options zfs zfs_arc_max=options zfs zfs_arc_max=8589934592` to `/etc/modprobe.d/zfs.conf`
2) I ran `sudo update-initramfs -u`
This worked well for a while (a few days) until I blindly accepted this update from Canonical..... note that Canonical is clearly fiddling with `initramfs` here. 

<img width="448" alt="Screenshot 2024-09-26 133730" src="https://github.com/user-attachments/assets/2e5a64ce-8b0b-497a-89fc-4b472e3ad2c4">

Once I accepted this update and rebooted, that killed ZFS and I got the `Failed to initialize the libzfs library` library when running `sudo zpool status`. 

Then: I was able to solve the issue by wiping the system, starting from scratch,  and NOT accepting this update from Canonical.

So I'm going to learn a lesson here:  "If you use ZFS on Ubuntu, if you think you have it stable, never update your system."

This may or may not be correct, but at least I can leave it here and see if anyone else stumbles over this and finds this thread. 

I would try to put this on Ubuntu's support forum ( https://ubuntuforums.org/ ) - but as far as I can tell, the search functionality of that forum is broken, it's very slow to load, and most of the commentary is very.... um... ANGRY..... 

Thanks.

**Reply by rincebrain** [reply]
That line is not valid to put in any modprobe config file, so I suspect there's a typo there, or that's your problem.
Lines in there look like
`options zfs foo=bar`
or sometimes
`blacklist foo`

**Reply by rincebrain** [reply]
Also, those lines are basically just telling you "initramfs-tools" and "initramfs-tools-core" got updated, which is not that interesting a result.

You should probably try loading the module, manually, after rebooting and it breaks, and seeing if it errors, because then that tells you that's a problem.

**Reply by kingpalethe** [reply]
Yikes, user error then likely. Closing discussion. thanks.
